aip/0133.md

```diff
@@ -138,7 +138,8 @@ message CreateBookRequest {
 
 - The `{resource}_id` field **must** exist on the request message, not the
   resource itself.
-- The `name` field on the resource **must** be ignored.
+- The `name` field on the resource **should** be empty or match the final
```

Comment:
This can be tricky due to interplay with the read-modify-write duplication interaction pattern. For example:

```python
book = GetBook("books/1")
book2 = deepcopy(book)
book2.title += ' (duplicate)'
CreateBook(id='duplicate of book', book2)
```

Not an amazing example, but nonetheless one that we have to keep in mind. If we don't ignore the book.name field, then it either needs to be updated or manually cleared out. And if the two differ for whatever reason, then we have to send back an error. 

It would also prevent us from annotating the "name" field as "Output only." -- output only fields must be ignored on input, so if we're not ignoring them, it's breaking that rule.

So while it might be confusing to have a world where you can create a resource and specify a parent, id, and a name on the resource that doesn't match either of those, the alternatives break some of the other assumptions and rules we have, so I'm -1 to this change.

Reply:
Hmm. My intuition is mixed here.

I agree that the behavior @ryanwe wants makes the `deepcopy` example work less well, but is that case more important than the lack of clarity involved in `create(id='x', book={'name': '.../y'})`? It _does_ seem a bit weird that we could take an invalid or conflicting value there and happily go along our merry way.

I am less sure how important either of these examples are in real life. I do agree that people probably do make a copy of a proto object and POST with it, but not sure how common that is. I also agree that people probably do muck up the id and name mismatch in other ways, where an `INVALID_ARGUMENT` might be nice, but I am not sure how common that is either.

Reply:
The important thing here is the Output-only distinction. If we do this, we have to either state that "name" is not an output only field or update AIP-203 to say that name is somehow special (since it says which says that we **cannot** throw an error for output only fields -- we must ignore the value (https://aip.dev/203#output-only)).

I'm of the opinion that name is an output only field. ID is an input-only field (implicitly, not explicitly). If that's the case, we **must** ignore name.

Reply:
Thanks for the interesting conversation.

I hadn't considered the read-modify-write-duplication scenario and that makes for a compelling argument to ignore the name. However, if we take it a step further and add an update, I feel this can be error prone. Building on your example:

```
book = GetBook("books/1")
book2 = deepcopy(book)
book2.title += ' (duplicate)'
CreateBook(id='duplicate of book', book2)

book2.author = 'new author'
UpdateBook(book2) // Updates the wrong book!!
```

Now the user has updated the wrong resource. Maybe this is acceptable since users would be expected to write `book2 = CreateBook(id='duplicate of book', book2)`. But (depending on the language) it seems like an easy mistake to make. And requiring users to update or clear the name, while adding some overhead, prevents these sorts of mistakes.

I may be mistaken, but I can imagine this happening in practice when 'deepcopy' is writing the response to disk and managing the resource as a json/yaml file. If this long-lived local version of the resource has the wrong name, it could lead to problems in the future.

Also, is name truly "output only" if it's a critical input field for updates? Granted that most use cases will follow a read-edit-update pattern, but that's not always the case, like when files or field masks are being used which requires users to manually set the name. Though maybe that puts it in the category of "immutable" fields?

So I think it comes down to whether in light of how Update() uses `name`, is it really output only? And do we optimize the read-modify-write-duplication scenario over ensuring correctness and avoiding potential confusion?

Reply:
There is another nuance to this when it comes to update, which is that for REST/JSON (still 75% of our users, I think), the name is auto-populated from the URL, e.g.

```
PATCH /v1/shelves/foo/books/bar
```

^-- automatically populates `book.name` with `shelves/foo/books/bar`.

It is possible in this case to send invalid data:

```
PATCH /v1/shelves/foo/books/bar

{"name": "shelves/baz/books/bar"}
```

...but to the best of my knowledge, ESF will just plow over it (meaning that for all intents and purposes, `name` _must_ be output only).

I think I agree that it is possible to get into bad situations here, but I think I agree with @jgeewax that the _status quo_ is still the best answer.